internal/provider/resource_access_offer.go

```diff
@@ -0,0 +1,577 @@
+	if len(users[permission.ReadAccess]) > 0 {
+		readUsersSet, errDiag := basetypes.NewSetValueFrom(ctx, types.StringType, users[permission.ReadAccess])
+		resp.Diagnostics.Append(errDiag...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		state.ReadUsers = readUsersSet
+	}
+	// Set the plan onto the Terraform state
+	state.OfferURL = types.StringValue(offerURL)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+// Update attempts to update the access to the offer.
+func (a *accessOfferResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "access offer", "update")
+		return
+	}
+	var plan, state accessOfferResourceOffer
+
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get the users to grant admin
+	var adminUsers []string
+	if !plan.AdminUsers.IsNull() {
+		resp.Diagnostics.Append(plan.AdminUsers.ElementsAs(ctx, &adminUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	// Get the users to grant consume
+	var consumeUsers []string
+	if !plan.ConsumeUsers.IsNull() {
+		resp.Diagnostics.Append(plan.ConsumeUsers.ElementsAs(ctx, &consumeUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	// Get the users to grant read
+	var readUsers []string
+	if !plan.ReadUsers.IsNull() {
+		resp.Diagnostics.Append(plan.ReadUsers.ElementsAs(ctx, &readUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	// validate if there are overlaps or admin user
+	// validation is done here considering dynamic (juju_user resource) and static values for users
+	err := validateNoOverlapsNoAdmin(adminUsers, consumeUsers, readUsers)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create access offer resource, got error: %s", err))
+		return
+	}
+
+	// If bob is in the plan but not in state = grant
+	// if bob is in the state but not in the plan = revoke
+	//    scenario 1 (users added/moved): bob was moved from admin to consume
+	//                bob will be 're-granted' consume permission in further steps
+	//    scenario 2 (users removed): bob was removed from the resource (user was in state but not in the plan anymore)
+	//                bob's permission will be revoked (revoke read) in the last update step
+
+	// scenario 1 - check for users added or moved
+	var readStateUsers []string
+	if !state.ReadUsers.IsNull() {
+		resp.Diagnostics.Append(state.ReadUsers.ElementsAs(ctx, &readStateUsers, false)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+	err = processPermissionChanges(plan.OfferURL.ValueString(), string(permission.ReadAccess), readUsers, readStateUsers, a.client)
```

Comment:
this is one way to  process permission changes, yes, but it relies heavily on the order in which we process these changes - first for the read access, consume access and finally for the admin access. any other order might result in incorrect permissions for users, i think.

one other way would be to simply revoke read access for all users in the state (this would remove all access rights for previous users) and then grant access per plan for each of the levels - this would mean slightly more churn, but would be order independent, i think.. wdyt?

Reply:
Agree. I used this approach considering the way that the old resource was defined to prevent changing existing permissions. Now, the update function revokes all state users and grants permissions accordingly to the plan.